tests/app/shared/comment-editor/comment-editor.spec.ts

```diff
@@ -0,0 +1,137 @@
+    const testUrl = 'testurl.com/test';
+
+    beforeEach(async(() => {
+      TestBed.configureTestingModule({
+        imports: [SharedModule, MarkdownModule.forChild(), BrowserAnimationsModule],
+        declarations: [CommentEditorComponent],
+        providers: [
+          { provide: UploadService, useValue: uploadService },
+          { provide: ErrorHandlingService, useValue: errorHandlingService },
+          { provide: LoggingService, useValue: logger }
+        ]
+      }).compileComponents();
+    }));
+
+    beforeEach(() => {
+      fixture = TestBed.createComponent(CommentEditorComponent);
+      form = new FormGroup({
+        description: new FormControl('')
+      });
+      commentField = form.get('description');
+      commentEditorComponent = fixture.componentInstance;
+      // Set up compulsory input fields
+      commentEditorComponent.commentField = commentField;
+      commentEditorComponent.commentForm = form;
+      commentEditorComponent.id = 'description';
+
+      commentTextArea = commentEditorComponent.commentTextArea;
+      textAreaEl = commentTextArea.nativeElement;
+      fixture.detectChanges();
+    });
+
+    it('should set up correctly', () => {
+      expect(fixture).toBeDefined();
+      expect(commentTextArea).toBeDefined();
+      expect(commentField).toBeDefined();
+    });
+
+    describe('insertUploadingText', () => {
+      describe('should set the correct value in the commentField', () => {
+        it('should insert a ! for displayable files', () => {
+          const expected = uploadTextTemplate(testDisplayableFilename);
+          insertUploadingText(testDisplayableFilename, commentField, commentTextArea);
+          expect(commentField.value).toBe(expected);
+        });
+
+        it('should not insert a ! for non-displayable files', () => {
+          const expected = uploadTextTemplate(testFilename);
+          insertUploadingText(testFilename, commentField, commentTextArea);
+          expect(commentField.value).toBe(expected);
+        });
+      });
+
+      it("should reposition the cursor by the uploading text's length", () => {
+        const initialPosition = textAreaEl.selectionEnd;
+        const expected = uploadTextTemplate(testDisplayableFilename);
+        insertUploadingText(testDisplayableFilename, commentField, commentTextArea);
+        expect(textAreaEl.selectionEnd).toBe(expected.length + initialPosition);
+      });
+    });
+
+    describe('insertUploadUrl', () => {
+      beforeEach(() => {
+        insertUploadingText(testDisplayableFilename, commentField, commentTextArea);
+      });
+
+      it('should replace the uploading text with the filename and link', () => {
+        insertUploadUrl(testDisplayableFilename, testUrl, commentField, commentTextArea);
+        const expected = uploadUrlTemplate(testDisplayableFilename, testUrl);
+        expect(commentField.value).toBe(expected);
+      });
+
+      describe('should position the cursor', () => {
```

Comment:
I understand your point as positioning cursor and also replacing text should actually be tested under `replacePlaceholderString`. 

I thought that a good way to test for `insertUploadUrl` and `insertUploadUrlVideo` would just actually be to spy on `replacePlaceholderString` and check that it was called with the correct arguments.

However, on digging into this approach, I found that as described in https://github.com/jasmine/jasmine/issues/1414, it is actually quite tricky to spy on exported functions due to how exported modules are going in the direction of being immutable (as mentioned [here](https://github.com/jasmine/jasmine/issues/1414#issuecomment-874241417). As such, my last commit does not test for text-replacement functionality for `replacePlaceholderString` directly, but through `insertUploadUrl` and `insertUploadUrlVideo`, which I don't think is ideal. I think a possible fix to this may be to refactor `upload-text-insertor` into a class instead of a module of exported functions, though I am not too sure about this.

Hope to hear your thoughts/thoughts from other reviewers, thanks a lot for looking through my PR!